File: libraries/pubsub.py
from docstore import connect as docstore_connect
from pika import BlockingConnection, ConnectionParameters
from re import match
from time import time as now
from urllib2 import build_opener, HTTPHandler, Request, urlopen

EXCHANGE = 'alexandra'
HTTP_PATIENCE_SEC = 1

class AMQPConnection:
    def __init__(self, url, context, marshal, unmarshal):
        self._context = context
        self.marshal, self.unmarshal = marshal, unmarshal
        host = match(r"amqp://([\w\d\.]+)", url).group(1)
        connection = BlockingConnection(ConnectionParameters(host))
        self._channel = connection.channel()
        self._channel.exchange_declare(exchange=EXCHANGE, type='topic')
        self._init_docstore()

    def _init_docstore(self):
        location_queue = self._subscribe_raw('docstore', 'location')
        self._publish_raw('docstore', 'locate', 'locate')
        docstore_url = self._get_message_block_raw(location_queue, timeout=1)
        self._docstore = docstore_connect(docstore_url)

    def publish(self, topic, message):
        self._publish_raw(self._context, topic, self.marshal(message))

    def _publish_raw(self, context, topic, message):
        self._channel.basic_publish(exchange=EXCHANGE,
                                    routing_key=context + '.' + topic,
                                    body=message)


    def subscribe(self, topic):
        return self._subscribe_raw(self._context, topic)

    def _subscribe_raw(self, context, topic):
        result = self._channel.queue_declare()
        queue = result.method.queue
        self._channel.queue_bind(exchange=EXCHANGE,
                                 queue=queue,
                                 routing_key=context + '.' + topic)
        return queue

    def unsubscribe(self, queue):
        self._channel.queue_delete(callback=None, queue=queue)

    def consume_queue(self, queue, f):
        def callback(ch, method, properties, body):
            f(self.unmarshal(body))

        self._channel.basic_consume(callback, queue=queue, no_ack=True)
        self._channel.start_consuming()

    def consume_topic(self, topic, f):
        queue = self.subscribe(topic)
        self.consume_queue(queue, f)

    def consume_all(self, f):
        queue = self.subscribe('#')
        def callback(ch, method, properties, body):
            amqp_topic = method.routing_key
            context, topic = amqp_topic.split('.', 1)
            f(context, topic, self.unmarshal(body))

        self._channel.basic_consume(callback, queue=queue, no_ack=True)
        self._channel.start_consuming()

    def get_message(self, queue):
        raw_message = self._get_message_raw(queue)
        if raw_message is None:
            return None
        else:
            return self.unmarshal(raw_message)

    def _get_message_raw(self, queue):
        return self._channel.basic_get(0, queue=queue, no_ack=True)[2]

    def get_all_messages(self, queue):
        messages = []
        while True:
            message = self.get_message(queue)
            if message is None:
                return messages
            else:
                messages.append(message)

    def get_message_block(self, queue, timeout=None):
        return self._get_message_block(queue, self.get_message, timeout)

    def _get_message_block_raw(self, queue, timeout=None):
        return self._get_message_block(queue, self._get_message_raw, timeout)

    def _get_message_block(self, queue, fetcher, timeout):
        alarm = Alarm(timeout)
        while True:
            message = fetcher(queue)
            if message is not None:
                return message
            if alarm.is_ringing():
                return None

    def get_current_message(self, topic):
        raw_message = self._docstore.get('/%s/%s' % (self._context, topic))
        if raw_message is None:
            return None
        else:
            return self.unmarshal(raw_message)

    def make_topic_monitor(self, topic):
        return TopicMonitor(self, topic)

class HTTPConnection:
    def __init__(self, url, context, marshal, unmarshal):
        self._root_url = '%s/contexts/%s' % (url, context)
        self.marshal, self.unmarshal = marshal, unmarshal

    def publish(self, topic, message):
        url = '%s/%s' % (self._root_url, topic)
        self._visit_url(url=url, data=self.marshal(message), method='POST')

    def subscribe(self, topic):
        return self._visit_url('%s/%s' % (self._root_url, topic))

    def unsubscribe(self, queue):
        url = '%s/queues/%s' % (self._root_url, queue)
        self._visit_url(url=url, method='DELETE')

    def consume_queue(self, queue, f):
        url = '%s/queues/%s' % (self._root_url, queue)
        headers = [('Patience', HTTP_PATIENCE_SEC)]
        while True:
            message = self._visit_url(url=url, headers=headers)
            if len(message) > 0:
                f(self.unmarshal(message))

    def consume_topic(self, topic, f):
        queue = self.subscribe(topic)
        self.consume_queue(queue, f)

    def consume_all(self, f):
        pass #not implemented

    def get_message(self, queue):
        url = '%s/queues/%s' % (self._root_url, queue)
        message = self._visit_url(url)
        if len(message) == 0:
            return None
        else:
            return self.unmarshal(message)

    def get_all_messages(self, queue):
        url = '%s/queues/%s' % (self._root_url, queue)
        headers = [('Range', 'all')]
        result = self._visit_url(url=url, headers=headers)
        if len(result) == 0:
            return []
        else:
            return map(self.unmarshal, result.split('\n'))

    def get_message_block(self, queue, timeout=None):
        url = '%s/queues/%s' % (self._root_url, queue)
        headers = [('Patience', HTTP_PATIENCE_SEC)]
        alarm = Alarm(timeout)
        while True:
            message = self._visit_url(url=url, headers=headers)
            if len(message) > 0:
                return self.unmarshal(message)
            if alarm.is_ringing():
                return None

    def make_topic_monitor(self, topic):
        return TopicMonitor(self, topic)

    def _visit_url(self, url, data=None, method='GET', headers=[]):
        opener = build_opener(HTTPHandler)
        request = Request(url)
        request.get_method = lambda: method
        for header in headers:
            request.add_header(*header)
        return opener.open(request, data).read()

connection_classes = {'amqp': AMQPConnection, 'http': HTTPConnection}
def identity(x):
    return x

def connect(url, context, marshal=identity, unmarshal=identity):
    protocol = match(r"(\w+)://", url).group(1)
    return connection_classes[protocol](url, context, marshal, unmarshal)

def firehose(url):
    return connect(url, '#')

class Alarm:
    def __init__(self, duration):
        if duration is not None:
            self.alarm_time = now() + duration
        else:
            self.alarm_time = None

    def is_ringing(self):
        return self.alarm_time is not None and now() > self.alarm_time

class TopicMonitor:
    def __init__(self, connection, topic):
        self._connection = connection
        self._queue = connection.subscribe(topic)
        self._latest = None

    def latest(self):
        messages = self._connection.get_all_messages(self._queue)
        if len(messages) > 0:
            self._latest = messages[-1]
        return self._latest

File: libraries/pubsub.rb
require 'bunny'
require './docstore.rb'
require 'net/http'

EXCHANGE = 'alexandra'
HTTP_PATIENCE_SEC = 1

module Pubsub

  class AMQPConnection
    def initialize(url, context, marshal, unmarshal)
      @context = context
      @marshal, @unmarshal = marshal, unmarshal
      host = %r{amqp://([\w\d\.]+)}.match(url)[1]
      conn = Bunny.new(:hostname => host)
      conn.start
      @channel = conn.create_channel
      @exchange = @channel.topic(EXCHANGE)
      init_docstore()
    end

    def init_docstore()
      location_queue = subscribe_raw('docstore', 'location')
      publish_raw('docstore', 'locate', 'locate')
      docstore_url = get_message_block_raw(location_queue, timeout=1)
      @docstore = Docstore::connect(docstore_url)
    end

    def publish(topic, message)
      publish_raw(@context, topic, @marshal.call(message))
    end

    def publish_raw(context, topic, message)
      @exchange.publish(message,
                        :routing_key => context + '.' + topic)
    end

    def subscribe(topic)
      return subscribe_raw(@context, topic)
    end

    def subscribe_raw(context, topic)
      queue = @channel.queue("")
      queue.bind(@exchange, :routing_key => context + '.' + topic)
      return queue
    end

    def unsubscribe(queue)
      queue.delete
    end

    def consume_queue(queue, &block)
      queue.subscribe(:block => true) do |di, p, body|
        block.call(@unmarshal.call(body))
      end
    end

    def consume_topic(topic, &block)
      queue = subscribe(topic)
      consume_queue(queue, &block)
    end

    def consume_all(&block)
      return nil # not implemented
    end

    def get_message(queue)
      if nil != get_message_raw(queue)
        @unmarshal.call(raw_message)
      else
        nil
      end
    end

    def get_message_raw(queue)
      return queue.pop[2]
    end

    def get_all_messages(queue)
      messages = []
      loop do
        message = get_message(queue)
        break if message == nil
        messages.push message
      end
      return messages
    end

    def get_message_block(queue, timeout=nil)
      get_message_block_internal(queue, timeout, method(:get_message))
    end

    def get_message_block_raw(queue, timeout=nil)
      get_message_block_internal(queue, timeout, method(:get_message_raw))
    end

    def get_message_block_internal(queue, timeout, fetcher)
      alarm = Alarm.new(timeout)
      loop do
        message = fetcher.call(queue)
        if message != nil
          return message
        elsif alarm.is_ringing()
          return nil
        end
      end
    end

    def get_current_message(topic)
      raw_message = @docstore.get('/%s/%s' % [@context, topic])
      if raw_message == nil
        return nil
      else
        return @unmarshal.call(raw_message)
      end
    end

    def make_topic_monitor(topic)
      return nil # not implemented
    end
  end

  class HTTPConnection
    def initialize(url, context, marshal, unmarshal)
      m = %r{http://([\w\d\.]+):(\d+)}.match(url)
      host, port = m[1], m[2].to_i
      @http = Net::HTTP.new(host, port)
      @prefix = '/contexts/%s' % context
      @marshal, @unmarshal = marshal, unmarshal
    end

    def publish(topic, message)
      path = '%s/%s' % [@prefix, topic]
      @http.send_request('POST', path, @marshal.call(message))
    end

    def subscribe(topic)
      response = @http.get('%s/%s' % [@prefix, topic])
      if response.code == "200"
        response.body
      else
        nil
      end
    end

    def consume_queue(queue, &block)
      path = '%s/queues/%s' % [@prefix, queue]
      loop do
        response = @http.get(path, {'Patience' => HTTP_PATIENCE_SEC.to_s})
        if response.code == "200" and response.body.length > 0
          block.call(@unmarshal.call(response.body))
        end
      end
    end

    def consume_topic(topic, &block)
      queue = subscribe(topic)
      consume_queue(queue, &block)
    end

    def consume_all(&block)
      return nil # not implemented
    end

    def unsubscribe(topic)
      path = '%s/queues/%s' % [@prefix, queue]
      @http.send_request('DELETE', path)
    end

    def get_message(queue)
      response = @http.get('%s/queues/%s' % [@prefix, queue])
      if response.code == "200" and response.body.length > 0
        @unmarshal.call(response.body)
      else
        nil
      end
    end

    def get_all_messages(queue)
      path = '%s/queues/%s' % [@prefix, queue]
      response = @http.get(path, {'Range' => 'all'})
      if response.code == "200" and response.body.length > 0
        messages = response.body.split("\n")
        messages.map(&@unmarshal)
      else
        nil
      end
    end

    def get_message_block(queue, timeout)
      path = '%s/queues/%s' % [@prefix, queue]
      alarm = Alarm.new(timeout)
      loop do
        response = @http.get(path, {'Patience' => HTTP_PATIENCE_SEC.to_s})
        if response.code == "200" and response.body.length > 0
          return @unmarshal.call(response.body)
        elsif alarm.is_ringing()
          return nil
        end
      end
    end

    def get_current_message(topic)
      response = @http.get('%s/%s' % [@prefix, topic], {'Range' => 'current'})
      if response.code == "200" and response.body.length > 0
        return @unmarshal.call(response.body)
      else
        return nil
      end
    end

    def make_topic_monitor(topic)
      return nil # not implemented
    end
  end

  def Pubsub.connect(url, context, marshal=-> x {x}, unmarshal=-> x {x})
    connection_classes = {'amqp' => AMQPConnection, 'http' => HTTPConnection}
    protocol = %r{(\w+)://}.match(url)[1]
    conn_class = connection_classes[protocol]
    return conn_class.new(url, context, marshal, unmarshal)
  end

  def Pubsub.firehose(url)
    return nil # not implemented
  end

  class Alarm
    def initialize(duration=nil)
      if duration != nil
        @alarm_time = Time.now + duration
      else
        @alarm_time = nil
      end
    end

    def is_ringing()
      @alarm_time != nil and Time.now > @alarm_time
    end
  end

  class QueueMonitor
    def initialize(connection, queue)
      @connection = connection
      @queue = queue
      @latest = nil
    end

    def latest()
      messages = @connection.get_all_messages(@queue)
      if not messages.empty?
        @latest = messages.last
      end
      return @latest
    end
  end
end
